File: LinkedLists/LinkedList.cpp
#include "LinkedList.h"
#include "Node.h"
#include "Node.cpp"
#include <stdio.h>
#include <iostream>
using namespace std;

LinkedList::LinkedList(int value){
    head->value = value;
}


void LinkedList::addFirst(int value){
    Node *temp = head;
    Node *newhead = new Node();
    newhead->value = value;
    newhead->next = temp;
    head = newhead;
}

void LinkedList::addLast(int val){
    Node *current = head;
    Node *nodeVal = new Node();
    nodeVal->value = val;
    while(current->next!=NULL){
        current = current->next; 
    }
    current->next = nodeVal;
}

// bool LinkedList::contains(int value){
//     bool isFound=false;

//     return false;
// }
// int LinkedList::indexOf(int value){

// }

void LinkedList::printList(){
    Node *current = head;
    while(current->next!=NULL){
        cout << current->value << endl;
        current = current->next; 
        if(current->next==NULL){
            cout << current->value << endl;
        }
    }
}

File: LinkedLists/src/LinkedList.java
import java.util.Arrays;

public class LinkedList {
    private class Node{
        private int value;
        private Node next;

        public Node(int value){
            this.value = value;
        }
    }

    private Node head, tail;
    private int size;

    public void addLast(int value){
        Node node = new Node(value);
        if (head==null) head = tail = node;
        else {
            tail.next = node;
            tail = node;
        }
        size++;
    }

    public void addFirst(int value){
        Node node = new Node(value);
        node.next = head;
        head = node;
        size++;
    }

    public void removeFirst(){
        Node temp = head.next;
        head = temp;
        size--;
    }

    public void removeLast(){
        Node current = head;
        while(current != null){
            if (current.next.next == null){
                tail = current;
                current.next = null;
            }
            current = current.next;
        }
        size--;
    }

    public int indexOf(int value){
        Node current = head;
        int i=0;
        while(current != null){
            if (current.value == value) return i;
            current = current.next;
            i++;
        }
        return -1;
    }
    public boolean contains(int value){
        if (indexOf(value) == -1) return false;
        else return true;
    }

    public void reverse(){
        Node prev = head, current = head.next;
        while(current != null){
            Node next_ = current.next;
            current.next = prev;
            prev = current;
            current = next_;
        }
        tail = head;
        tail.next = null;
        head = prev;
    }

    
    
    public int size(){
        return this.size;
    }

    public int[] returnArray(){
        int[] arr = new int[size];
        Node current = head;
        int i=0;
        while(current != null){
            // System.out.println(current.value);
            arr[i++] = current.value;
            current = current.next;
        }
        // System.out.println(arr.toString());
        return arr;
    }

    public void print(){
        int[] arr = returnArray();
        System.out.println(Arrays.toString(arr));
    }
}
